### PR TITLE
Feature/notification pop up

### PR DESCRIPTION
**Pull request by programmerkenzi**
# Description

notification interface and class, with added functionality to save and remove data to local storage.

**version notification  & press btn to close it**

notifiy

**reward notification:**

截圖 2023-03-23 下午2 45 28

Fixes # (issue)

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] Test A
- [ ] Test B

**Test Configuration**:
* Firmware version:
* Hardware:
* Toolchain:
* SDK:

# Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: src/components/Notification.tsx [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+import { INotification } from '~/types/notification'
+
+import Popup from './Popup'
+const Notification = ({
+    id,
+    title,
+    subtitle,
+    contents,
+    img,
+}: INotification) => {
```
Comment (ahwei) [inline src/components/Notification.tsx]: useEffect here if need to show

File: src/container/Home/index.tsx [hunk 2]
```diff
@@ -110,8 +112,14 @@ const Home = ({ initialTab = 0 }: IProps) => {
 
     useEffect(() => {
         getExchangeRate()
+        if (!locked) {
```
Comment (ahwei) [inline src/container/Home/index.tsx]: remove it and use `<Notification/>`

File: src/store/NotificationStore.tsx [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+            id: `notification-${notificationId}`,
+        }
+
+        const data = [...this.notification, notification]
+        console.log(
+            '🚀 ~ file: NotificationStore.tsx:43 ~ NotificationManager ~ add ~ data:',
+            data
+        )
+        this.notification = [...data]
+        setStorage({ notification: [...data] })
```
Comment (ahwei) [inline src/store/NotificationStore.tsx]: move to Notification

File: src/store/NotificationStore.tsx [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+
+    public add(options: VersionNotification | TxRewardNotification): void {
+        const notificationId = uuidv4()
+        const notification: INotification = {
+            ...options, // if id is passed within options, it will overwrite the auto-generated one
+            id: `notification-${notificationId}`,
+        }
+
+        const data = [...this.notification, notification]
+        console.log(
```
Comment (ahwei) [inline src/store/NotificationStore.tsx]: remove  thx

File: src/types/notification.ts [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+    INotification,
+    'title' | 'subtitle' | 'contents'
+>
+
+export type TxRewardNotification = Pick<
+    INotification,
+    'img' | 'title' | 'subtitle'
+>
+
+export const versionData: VersionNotification = {
```
Comment (ahwei) [inline src/types/notification.ts]: try to move to public

File: src/components/Popup.tsx [hunk 6]
```diff
@@ -2,41 +2,73 @@ import classNames from 'classnames'
...
 import { useTranslation } from 'react-i18next'
 
+import CloseIcon from '~icons/hoogii/close.jsx'
+
 interface IProps {
+    id?: string
     actionButton?: ReactNode
     close?: () => void
     className?: string
+    childrenClassName?: string
```
Comment (ahwei) [inline src/components/Popup.tsx]: use  `actionButton?: ReactNode` to solve it

File: src/components/Popup.tsx [hunk 7]
```diff
@@ -2,41 +2,73 @@ import classNames from 'classnames'
...
     children,
+    childrenClassName,
+    closeBtn = true,
+    btnClassName,
+    closeIconBtn = false,
 }: PropsWithChildren<IProps>) => {
     const { t } = useTranslation()
     return (
-        <div className="fixed inset-0 z-50 px-5 py-auto flex-center">
+        <div className="fixed inset-0 z-50 px-5 py-auto flex-center" id={id}>
```
Comment (ahwei) [inline src/components/Popup.tsx]: add new component to solve popup

File: src/store/NotificationStore.tsx [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+        )
+        this.render()
+    }
+
+    public close(id: string): void {
+        console.log(
+            '🚀 ~ file: NotificationStore.tsx:49 ~ NotificationManager ~ close ~ id:',
+            id
+        )
+        const data = this.notification.filter(
```
Comment (ahwei) [inline src/store/NotificationStore.tsx]: if pkg.version === false .
note: use  `pkg.version.split('-')[0]`

File: src/components/Notification.tsx [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+
+        setData(() => newData)
+    }
+
+    const initData = async () => {
+        const localData =
+            (await getStorage<INotification[]>('notifications')) || []
+        setData(() => [...localData])
+        // check whether the version notification has been displayed
+        const curVersion = isDev ? pkg.version : pkg.version.split('-')[0]
```
Comment (ahwei) [inline src/components/Notification.tsx]: just use `pkg.version.split('-')[0]`, some times we build package is use version+dev.XXX .

File: src/index.scss [hunk 10]
```diff
@@ -151,6 +151,15 @@
             mask-composite: exclude;
         }
     }
+
+    .btn {
+        &-md {
+            &-primary {
+                @apply min-w-[100px] min-h-[28px]  text-center align-middle font-bold text-xs  bg-primary-500 rounded-[4px] text-white #{!important};
```
Comment (ahwei) [inline src/index.scss]: remove  #{!important} or move this style to `Notification.tsx` button

File: src/types/notification.ts [hunk 11]
```diff
@@ -0,0 +1,30 @@
...
+    INotification,
+    'title' | 'subtitle' | 'contents'
+>
+
+export type TxRewardNotification = Pick<
+    INotification,
+    'img' | 'title' | 'subtitle'
+>
+
+export const rewardData: TxRewardNotification = {
```
Comment (ahwei) [inline src/types/notification.ts]: this file is for type declare ,move to `Notification.tsx ` or remove it

File: src/components/Notification.tsx [hunk 12]
```diff
@@ -0,0 +1,122 @@
...
+
+    const handleOnClose = (version?: string) => {
+        if (version) {
+            setStorage({ last_version_notification: version })
+        }
+
+        const newData = [...data]
+        newData.shift()
+
+        setData(() => newData)
```
Comment (ahwei) [inline src/components/Notification.tsx]: you can try
```tsx
setData((data) =>data.shift())
```

File: src/components/Notification.tsx [hunk 13]
```diff
@@ -0,0 +1,122 @@
...
+    const initData = async () => {
+        const localData =
+            (await getStorage<INotification[]>('notifications')) || []
+        setData(() => [...localData])
+        // check whether the version notification has been displayed
+        const curVersion = isDev ? pkg.version : pkg.version.split('-')[0]
+        if (curVersion === versionData.version) {
+            const last_version_notification =
+                (await getStorage<string | boolean>(
+                    'last_version_notification'
```
Comment (ahwei) [inline src/components/Notification.tsx]: v1.0.7. v1.0.6 or null....
if curVersion > last_version_notification

File: src/components/Notification.tsx [hunk 14]
```diff
@@ -0,0 +1,122 @@
...
+    useEffect(() => {
+        initData()
+
+        // TODO: listen for websocket messages for other notifications
+    }, [])
+
+    return (
+        data.length > 0 && (
+            <div className="flex items-center justify-center absolute-full bg-overlay">
+                {data
```
Comment (ahwei) [inline src/components/Notification.tsx]: use `data[0]`

File: src/types/version.d.ts [hunk 15]
```diff
@@ -0,0 +1,6 @@
+import { VersionNotification } from './notification'
+
+declare module '*.json' {
```
Comment (ahwei) [inline src/types/version.d.ts]: try version.json

File: src/components/Notification.tsx (hunk 12 above)
Comment (programmerkenzi) [reply src/components/Notification.tsx]: data.shift() will return the item that remove , will use data.slice()

**Review by ahwei — APPROVED** [review]

**Pull request merged**